### PR TITLE
kubeadm: add new testcase to check the initConfiguration on worker nodes

### LINKED ISSUE #2943: failing kinder jobs at latest that call kubeadm reset
https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-external-ca-latest
Started to fail yesterday. @chendave @SataQiu @neolit123 

```

I1016 15:45:24.588701   94159 etcd.go:214] retrieving etcd endpoints from "kubeadm.kubernetes.io/etcd.advertise-client-urls" annotation in etcd Pods
E1016 15:45:24.589382   94159 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1bbd500?, 0x3116220})
	vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x1e29dc0?})
	vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x1bbd500?, 0x3116220?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
k8s.io/client-go/kubernetes.(*Clientset).CoreV1(0xc000692d00?)
	vendor/k8s.io/client-go/kubernetes/clientset.go:306
k8s.io/kubernetes/cmd/kubeadm/app/util/etcd.getRawEtcdEndpointsFromPodAnnotationWithoutRetry({0x21a2050, 0x0})
	cmd/kubeadm/app/util/etcd/etcd.go:215 +0xde
k8s.io/kubernetes/cmd/kubeadm/app/util/etcd.getRawEtcdEndpointsFromPodAnnotation.func1()
	cmd/kubeadm/app/util/etcd/etcd.go:190 +0x3c
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x36?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:145 +0x3e
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x3b9aca00, 0x3ff0000000000000, 0x3fb999999999999a, 0x1e, 0x0}, 0xc0000c47e0?)
	vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:461 +0x5a
k8s.io/kubernetes/cmd/kubeadm/app/util/etcd.getRawEtcdEndpointsFromPodAnnotation({0x21a2050?, 0x0?}, {0x3b9aca00, 0x3ff0000000000000, 0x3fb999999999999a, 0x1e, 0x0})
	cmd/kubeadm/app/util/etcd/etcd.go:188 +0xd9
k8s.io/kubernetes/cmd/kubeadm/app/util/etcd.getEtcdEndpointsWithBackoff(...)
	cmd/kubeadm/app/util/etcd/etcd.go:179
k8s.io/kubernetes/cmd/kubeadm/app/util/etcd.getEtcdEndpoints({0x21a2050?, 0x0?})
	cmd/kubeadm/app/util/etcd/etcd.go:175 +0x5c
k8s.io/kubernetes/cmd/kubeadm/app/util/etcd.NewFromCluster({0x21a2050?, 0x0?}, {0x0, 0x0})
	cmd/kubeadm/app/util/etcd/etcd.go:146 +0x45
k8s.io/kubernetes/cmd/kubeadm/app/phases/etcd.RemoveStackedEtcdMemberFromCluster({0x21a2050, 0x0}, 0xc0004b0e00)
	cmd/kubeadm/app/phases/etcd/local.go:94 +0xd9
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/reset.runRemoveETCDMemberPhase({0x1d40aa0?, 0xc00038f000?})
	cmd/kubeadm/app/cmd/phases/reset/removeetcdmember.go:66 +0x28a
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1(0xc0003518c0)
	cmd/kubeadm/app/cmd/phases/workflow/runner.go:259 +0x11a
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll(...)
	cmd/kubeadm/app/cmd/phases/workflow/runner.go:446
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run(0xc00068e3f0, {0xc000390780, 0x0, 0x2})
	cmd/kubeadm/app/cmd/phases/workflow/runner.go:232 +0x3b9
k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdReset.func1(0xc00034c800?, {0xc000390780, 0x0, 0x2})
	cmd/kubeadm/app/cmd/reset.go:214 +0x65
github.com/spf13/cobra.(*Command).execute(0xc00034e300, {0xc000390760, 0x2, 0x2})
	vendor/github.com/spf13/cobra/command.go:940 +0x87c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001dfb00)
	vendor/github.com/spf13/cobra/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(0x21672e0?)
	vendor/github.com/spf13/cobra/command.go:992 +0x13
k8s.io/kubernetes/cmd/kubeadm/app.Run()
	cmd/kubeadm/app/kubeadm.go:50 +0x155
main.main()
	cmd/kubeadm/kubeadm.go:25 +0x13
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xc0 pc=0x14eade0]
```

### PR DESCRIPTION
**Pull request by chendave**
etcd should not be removed on worker node, and this is checked by if the `initCfg` is `nil`.

`admin.conf` doesn't exsit on workers, this will make sure The `initCfg` stay as `nil`.

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubeadm/issues/2943
ref: https://github.com/kubernetes/kubernetes/pull/119946

#### Special notes for your reviewer:
https://github.com/kubernetes/kubernetes/blob/86676bce04dac096f89bea4d0fcbc6452630c5d2/cmd/kubeadm/app/cmd/phases/reset/removeetcdmember.go#L64-L67

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/reset.go [hunk 1]
```diff
@@ -105,7 +105,7 @@ func newResetData(cmd *cobra.Command, opts *resetOptions, in io.Reader, out io.W
 		return nil, err
 	}
 
-	initCfg := &kubeadmapi.InitConfiguration{}
+	var initCfg *kubeadmapi.InitConfiguration
```
Comment (SataQiu) [inline cmd/kubeadm/app/cmd/reset.go:108]: Add some comment to avoid same wrong refactor?
It would be nice to add some unit tests about this.

Comment (chendave) [reply cmd/kubeadm/app/cmd/reset.go:109]: Done

**Comment by chendave** [comment]
this is ready for review now.

File: cmd/kubeadm/app/cmd/reset.go [hunk 2]
```diff
@@ -127,7 +128,11 @@ func newResetData(cmd *cobra.Command, opts *resetOptions, in io.Reader, out io.W
 		klog.V(1).Infof("[reset] Could not obtain a client set from the kubeconfig file: %s", opts.kubeconfigPath)
 	}
 
-	ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(initCfg.FeatureGates, opts.ignorePreflightErrors, resetCfg.IgnorePreflightErrors)
+	featureList := make(map[string]bool)
+	if initCfg != nil {
+		featureList = initCfg.FeatureGates
+	}
```
Comment (SataQiu) [inline cmd/kubeadm/app/cmd/reset.go:134]: I'm a little confused. User should how to enable `MergeCLIArgumentsWithConfig` feature gate for `kubeadm reset`? It seems that the `featureList` will always be empty on worker nodes. Or I have missed something.

File: cmd/kubeadm/app/cmd/reset_test.go [hunk 3]
```diff
@@ -29,6 +29,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta4"
 	kubeadmapiv1 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta4"
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/reset_test.go:33]: there is already a v1beta4 import here.

File: cmd/kubeadm/app/cmd/reset.go (hunk 2 above)
Comment (SataQiu) [reply cmd/kubeadm/app/cmd/reset.go:134]: I'm not sure if we handled feature gate correctly for `init/join/reset/upgrade`, we may need a double check for #119946 @chendave @pacoxu @neolit123

File: cmd/kubeadm/app/cmd/reset_test.go (hunk 3 above)
Comment (chendave) [reply cmd/kubeadm/app/cmd/reset_test.go]: Done

File: cmd/kubeadm/app/cmd/reset.go (hunk 2 above)
Comment (chendave) [reply cmd/kubeadm/app/cmd/reset.go:134]: If it's a worker node, the initCfg is nil, there is no way to tell if `MergeCLIArgumentsWithConfig` is enabled or not, the CLI override the config file is the default behavior.

Comment (chendave) [reply cmd/kubeadm/app/cmd/reset.go:134]: it's looks like to me this is a limitation of how FG is implemented in kubeadm, it's only applicable to control nodes, not just this FG but all other FG should be true as well.

Comment (chendave) [reply cmd/kubeadm/app/cmd/reset.go:134]: two choice here, 
- document this as a limitation, FG is now not configurable in the reset/join/init but only in the clusterCfg.
- or we need to revert https://github.com/kubernetes/kubernetes/pull/119946 and see if there is any other choice to implement it.

Comment (chendave) [reply cmd/kubeadm/app/cmd/reset.go:134]: I am fine with revert.

@neolit123 @pacoxu WDYT?

Comment (pacoxu) [reply cmd/kubeadm/app/cmd/reset.go:134]: +1

**Comment by chendave** [comment]
close this as we decide to revert this https://github.com/kubernetes/kubernetes/pull/119946 eventually.

**Comment by chendave** [comment]
testcase is still useful

**Comment by chendave** [comment]
/retitle kubeadm: add new testcase to check the initConfiguration on worker nodes

**Comment by neolit123** [comment]
> admin.conf doesn't exsit on workers, this will make sure The initCfg stay as nil.

/hold

like i said on the other ticket 
https://github.com/kubernetes/kubernetes/pull/121348#issuecomment-1770368486

kubeadm reset of workers should be able to still get the ClusterConfiguraiton and construct a best effort InitConfiguration.
did we break something or are my assumptions out of date / wrong?

**Comment by chendave** [comment]
> kubeadm reset of workers should be able to still get the ClusterConfiguraiton and construct a best effort InitConfiguration.

revisit https://github.com/kubernetes/kubernetes/pull/119946 is in my TODO list, the UT added here intends to prevent from CI failure. 

even we can construct a InitConfiguration, should us include that in the reset data? if yes, we should revisit the check here as well.
https://github.com/kubernetes/kubernetes/blob/86676bce04dac096f89bea4d0fcbc6452630c5d2/cmd/kubeadm/app/cmd/phases/reset/removeetcdmember.go#L64-L67

I think something did [here ](https://github.com/kubernetes/kubernetes/pull/119946), in a future pr, might not touch reset data, it looks like to me two different things.

**Comment by chendave** [comment]
> I think something did https://github.com/kubernetes/kubernetes/pull/119946, in a future pr, might not touch reset data,

my current thinking is that on a worker node, it will read from the configmap to get the FG information instead of from reset data.

**Comment by chendave** [comment]
since we are not clear what to do next for: https://github.com/kubernetes/kubernetes/pull/119946

lets close it for now, we can continue to discuss when the new PR is ready.

**Pull request closed without merging**